src/main/java/slimeknights/tconstruct/plugin/jei/ToolBuildingCategory.java

```diff
@@ -0,0 +1,147 @@
+import slimeknights.tconstruct.library.tools.layout.LayoutSlot;
+import slimeknights.tconstruct.library.tools.part.IToolPart;
+import slimeknights.tconstruct.tools.TinkerTools;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Stream;
+
+public class ToolBuildingCategory implements IRecipeCategory<ToolBuildingRecipe> {
+  private static final ResourceLocation BACKGROUND_LOC = TConstruct.getResource("textures/gui/jei/tinker_station.png");
+  private static final Component TITLE = TConstruct.makeTranslation("jei", "tinkering.tool_building");
+  @Getter
+  private final IDrawable icon;
+  @Getter
+  private final IDrawable background;
+  private final IDrawable anvil, slot;
+  private static final int WIDTH = 128;
+  private static final int HEIGHT = 66;
+  private static final int SLOT_SIZE = 18;
+  private static final int ITEM_SIZE = 16;
+
+  public ToolBuildingCategory(IGuiHelper guiHelper) {
+    this.icon = guiHelper.createDrawableItemStack(TinkerTools.pickaxe.get().getRenderTool());
+    this.background = guiHelper.createDrawable(BACKGROUND_LOC, 128, 77, WIDTH, HEIGHT);
+    this.slot = guiHelper.createDrawable(BACKGROUND_LOC, 24, 14, SLOT_SIZE, SLOT_SIZE);
+    this.anvil = guiHelper.createDrawable(BACKGROUND_LOC, 128, 61, ITEM_SIZE, ITEM_SIZE);
+  }
+
+  @Override
+  public void setRecipe(IRecipeLayoutBuilder builder, ToolBuildingRecipe recipe, IFocusGroup focuses) {
+    List<IToolPart> parts = recipe.getToolParts();
+    List<ItemStack> extraItems = recipe.getIngredients().stream().flatMap(ingredient -> Arrays.stream(ingredient.getItems())).toList();
+    List<LayoutSlot> layoutSlots = recipe.getLayoutSlots();
+
+    // every part and its material variants
+    List<ItemStack> allInputParts = parts.stream()
+        .flatMap(part -> MaterialRegistry.getInstance().getVisibleMaterials().stream()
+          .filter(part::canUseMaterial)
+          .flatMap(mat -> Stream.of(part.withMaterial(mat.getIdentifier()))))
+        .toList();
+
+    if (parts.size() + extraItems.size() != layoutSlots.size()) {
+      TConstruct.LOG.error(String.format("Tool part count and layout slot count for %s do not match!", recipe.getId()));
+      return;
+    }
+
+    builder.addInvisibleIngredients(RecipeIngredientRole.INPUT).addItemStacks(allInputParts);
+    builder.addInvisibleIngredients(RecipeIngredientRole.INPUT).addItemStacks(extraItems);
+
+    Vec2 offsets = getOffsets(layoutSlots);
+    for (int i = 0; i < parts.size(); i++) {
+      builder.addSlot(RecipeIngredientRole.INPUT, (int) (layoutSlots.get(i).getX() + offsets.x), (int) (layoutSlots.get(i).getY() + offsets.y))
+             .addItemStack(getDisplayPart(parts.get(i), i));
+    }
+
+    if (recipe.getOutput() instanceof IModifiableDisplay modifiable) {
+      builder.addSlot(RecipeIngredientRole.OUTPUT, WIDTH - 23, 23)
+        .addItemStack(modifiable.getRenderTool());
+    }
```

Comment:
This needs a fallback if that instanceof check fails. Probably just a `new ItemStack(recipe.getOutput())`